### ISSUE TITLE
torch.stack should have a pin_memory parameter

### ISSUE THREAD

**Issue by moi90** [pytorch/pytorch#69213]
## 🚀 Feature
torch.stack should have a pin_memory parameter: `torch.stack(tensors, dim=0, *, out=None, pin_memory=False)`

## Motivation

`torch.stack` "Concatenates a sequence of tensors along a new dimension." This usually means that data is copied from different locations to one memory location. After having assembled a batch of images with `torch.stack`, I want to copy it to the GPU as fast as possible. To enable non-blocking transfer, the tensor has to be in pinned memory. But to to that, the data has to be copied into pinned memory first. This is one copy too much.

## Pitch

`torch.stack` should have a `pin_memory` parameter so that the individual data pieces are copied directly into pinned memory.

## Alternatives

Currently, I'm using the following:

```python
def stack_ex(
    tensors: Union[Tuple["torch.Tensor"], List["torch.Tensor"]], pin_memory: bool = False
):
    n_tensors = len(tensors)
    first = tensors[0]
    size = (len(tensors),) + first.size()
    out = torch.empty(size, dtype=first.dtype, pin_memory=pin_memory)
    torch.stack(tensors, out=out)
    return out
```

But I don't know how efficient that is.

**Comment by moi90** [comment]
Any thoughts on this? Is it a good idea?
